### ISSUE TITLE
Client Credentials grant not working in 2.0.11

### ISSUE THREAD

**Issue by Damore** [spring-projects/spring-security-oauth#896]
After I updated the jar, I only got the following error when requesting a token: 

`{
  "error": "invalid_client",
  "error_description": "Bad client credentials"
}`

Before the update I was requesting a token to the same url "/oauth/token" and I was receiving the token normally:

`{
  "access_token": "c15d138a-681e-4e62-98c7-1440fa600686",
  "token_type": "bearer",
  "refresh_token": "d7ab807c-f884-4b5d-8013-5f0549d61583",
  "expires_in": 899,
  "scope": "read write"
}`

What should I do to make this work on version 2.0.11? This issue is related to this [question](http://stackoverflow.com/questions/40283365/spring-boot-oauth2-not-working-properly-on-version-1-4-1)

**Comment by wilkinsona** [comment]
The Spring Boot issue referenced above contains a sample that will reproduce the problem.

**Comment by jgrandja** [comment]
Hi @Damore. I was able to reproduce this issue but it took some effort to get setup. As @wilkinsona mentioned in this [comment](https://github.com/spring-projects/spring-boot/issues/7258#issuecomment-257353764), we do need a _**minimal sample**_ in order to efficiently troubleshoot.

In order for me to set this up, I had to add the missing dependency (Commons Lang 3) and comment out Spring Cloud Config deps as I don't have a config server setup. I also had to setup MongoDB. Going forward please supply a minimal sample reproducing the issue that can be setup quickly in any dev env (not just eclipse as I use IntelliJ).

I'm looking at this issue now...

**Comment by jgrandja** [comment]
Hi @Damore. A security vulnerability was reported via #808 and fixed in 2.0.11.

The issue you are having is the result of this fix.
However, your configuration of `CustomUserDetailsService` is not correct.

Your current configuration in `WebSecurityConfiguration` is as follows:

```
	@Override
	protected void configure(AuthenticationManagerBuilder auth) throws Exception {
		auth.userDetailsService(customUserDetailsService).passwordEncoder(passwordEncoder);
		super.configure(auth);
	}
```

You are configuring the `UserDetailsService` for the parent `AuthenticationManager` with your `CustomUserDetailsService` which is backed by your **Clients** **NOT** your **Users**. The parent `AuthenticationManager` should be configured with a `UserDetailsService` backed by your **User** database. This configuration needs to be removed.

I noticed the same configuration in `ResourceServer`

```
http.sessionManagement()......and().userDetailsService(customUserDetailsService);
```

This configuration needs to be removed as well.

**The following will fix your issue...**

Your current configuration in `OAuth2Config` is as follows:

```
	@Autowired
	private ClientDetailsService customClientDetailsService;
	
    @Override
    public void configure(AuthorizationServerSecurityConfigurer security) throws Exception {
	security.allowFormAuthenticationForClients();
    }

	@Override
	public void configure(ClientDetailsServiceConfigurer clients) throws Exception {
		clients.withClientDetails(customClientDetailsService);
	}
```

Change it to:

```
	@Autowired
	public PasswordEncoder passwordEncoder;

	@Autowired
	private ClientDetailsService customClientDetailsService;
	
    @Override
    public void configure(AuthorizationServerSecurityConfigurer security) throws Exception {
        security.allowFormAuthenticationForClients().passwordEncoder(passwordEncoder);
    }

	@Override
	public void configure(ClientDetailsServiceConfigurer clients) throws Exception {
		clients.withClientDetails(customClientDetailsService);
	}
```

FYI, the reason why you were getting _Bad Client Credentials_ is because the `PasswordEncoder` (BCrypt) was not correctly configured. When the password comparison occurred it failed because it wasn't decoded after it was fetched from MongoDB.

I'm going to close this issue as the solution documented above worked.
Thanks.
